### ISSUE TITLE
Make DynamicNUTS robust to numerical errors

### ISSUE THREAD

**Issue by itsdfish** [TuringLang/Turing.jl#767]
Samplers such as NUTS will continue to run when a model produces an occasional numerical error. Typically, a few of these errors is not problematic.  However, DynamicsNUTS stops running completely as shown below. I have two questions: (1) is the error handling on the Turing on DynamicHMC side, (2) is it possible to let DynamicNUTS print a warning instead and continue to run? 

```
ERROR: LoadError: InvalidLogDensityException: value is NaN
Stacktrace:
 [1] LogDensityProblems.ValueGradient{Float64,Array{Float64,1}}(::Float64, ::Array{Float64,1}) at /home/dfish/.julia/packages/ArgCheck/xX4DA/src/checks.jl:165
 [2] Type at /home/dfish/.julia/packages/LogDensityProblems/KQ9nc/src/result_types.jl:83 [inlined]
 [3] (::getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}})(::Array{Float64,1}) at /home/dfish/.julia/packages/Turing/meCJf/src/inference/dynamichmc.jl:62
 [4] logdensity at /home/dfish/.julia/packages/Turing/meCJf/src/Turing.jl:142 [inlined]
 [5] leapfrog(::DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Symmetric{Float64,Array{Float64,2}},LinearAlgebra.LowerTriangular{Float64,Array{Float64,2}}}}, ::DynamicHMC.PhasePoint{Array{Float64,1},LogDensityProblems.ValueGradient{Float64,Array{Float64,1}}}, ::Float64) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/hamiltonian.jl:190
 [6] move at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/buildingblocks.jl:315 [inlined]
 [7] adjacent_tree(::MersenneTwister, ::DynamicHMC.Trajectory{DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Symmetric{Float64,Array{Float64,2}},LinearAlgebra.LowerTriangular{Float64,Array{Float64,2}}}},Float64}, ::DynamicHMC.PhasePoint{Array{Float64,1},LogDensityProblems.ValueGradient{Float64,Array{Float64,1}}}, ::Int64, ::Bool) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/buildingblocks.jl:59
 [8] adjacent_tree(::MersenneTwister, ::DynamicHMC.Trajectory{DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Symmetric{Float64,Array{Float64,2}},LinearAlgebra.LowerTriangular{Float64,Array{Float64,2}}}},Float64}, ::DynamicHMC.PhasePoint{Array{Float64,1},LogDensityProblems.ValueGradient{Float64,Array{Float64,1}}}, ::Int64, ::Bool) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/buildingblocks.jl:65 (repeats 2 times)
 [9] sample_trajectory(::MersenneTwister, ::DynamicHMC.Trajectory{DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Symmetric{Float64,Array{Float64,2}},LinearAlgebra.LowerTriangular{Float64,Array{Float64,2}}}},Float64}, ::DynamicHMC.PhasePoint{Array{Float64,1},LogDensityProblems.ValueGradient{Float64,Array{Float64,1}}}, ::Int64) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/buildingblocks.jl:103
 [10] #NUTS_transition#14(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::MersenneTwister, ::DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Symmetric{Float64,Array{Float64,2}},LinearAlgebra.LowerTriangular{Float64,Array{Float64,2}}}}, ::Array{Float64,1}, ::Float64, ::Int64) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/buildingblocks.jl:366
 [11] NUTS_transition(::MersenneTwister, ::DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Symmetric{Float64,Array{Float64,2}},LinearAlgebra.LowerTriangular{Float64,Array{Float64,2}}}}, ::Array{Float64,1}, ::Float64, ::Int64) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/buildingblocks.jl:364
 [12] mcmc_adapting_ϵ(::DynamicHMC.NUTS{Array{Float64,1},Float64,MersenneTwister,DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Symmetric{Float64,Array{Float64,2}},LinearAlgebra.LowerTriangular{Float64,Array{Float64,2}}}},ReportIO{Base.TTY}}, ::Int64, ::DynamicHMC.DualAveragingParameters{Float64}, ::DynamicHMC.DualAveragingAdaptation{Float64}) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:73
 [13] mcmc_adapting_ϵ at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:83 [inlined]
 [14] tune(::DynamicHMC.NUTS{Array{Float64,1},Float64,MersenneTwister,DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Symmetric{Float64,Array{Float64,2}},LinearAlgebra.LowerTriangular{Float64,Array{Float64,2}}}},ReportIO{Base.TTY}}, ::DynamicHMC.StepsizeCovTuner{Float64}) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:211
 [15] tune(::DynamicHMC.NUTS{Array{Float64,1},Float64,MersenneTwister,DynamicHMC.Hamiltonian{Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}},GaussianKE{LinearAlgebra.Diagonal{Float64,Array{Float64,1}},LinearAlgebra.LowerTriangular{Float64,LinearAlgebra.Diagonal{Float64,Array{Float64,1}}}}},ReportIO{Base.TTY}}, ::DynamicHMC.TunerSequence{Tuple{DynamicHMC.StepsizeTuner,DynamicHMC.StepsizeCovTuner{Float64},DynamicHMC.StepsizeCovTuner{Float64},DynamicHMC.StepsizeCovTuner{Float64},DynamicHMC.StepsizeCovTuner{Float64},DynamicHMC.StepsizeCovTuner{Float64},DynamicHMC.StepsizeTuner}}) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:262
 [16] #NUTS_init_tune_mcmc#20(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::MersenneTwister, ::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}, ::Int64) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:285
 [17] NUTS_init_tune_mcmc at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:284 [inlined]
 [18] #NUTS_init_tune_mcmc#21 at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:294 [inlined]
 [19] NUTS_init_tune_mcmc(::Turing.Inference.FunctionLogDensity{getfield(Turing.Inference, Symbol("#_lp#49")){Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}},Turing.Sampler{DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}}}}, ::Int64) at /home/dfish/.julia/packages/DynamicHMC/gnaxv/src/sampler.jl:294
 [20] sample(::Turing.Model{Tuple{:τ,:A,:k,:υ1,:υ2},Tuple{:data},getfield(Main, Symbol("###inner_function#437#36")),NamedTuple{(:data,),Tuple{Array{Tuple{Int64,Float64},1}}},NamedTuple{(:data,),Tuple{Symbol}}}, ::DynamicNUTS{Turing.Core.ForwardDiffAD{40},Union{}}) at /home/dfish/.julia/packages/Turing/meCJf/src/inference/dynamichmc.jl:65
 [21] top-level scope at none:0
in expression starting at /home/dfish/Dropbox/Julia/Bayesian LBA/Run LBA Turing.jl:28
```

**Comment by yebai** [comment]
@itsdfish This is handled by `DynamicHMC` at the moment, and `DynamicHMC` throws an error when numerical errors are met. If you are interested, please take a look at:

https://github.com/tpapp/LogDensityProblems.jl/blob/master/src/reject_errors.jl

We are trying to handle numerical errors in a more user-agonistic way in `AdvancedHMC`. The plan is to make it available in release 0.7.

**Comment by itsdfish** [comment]
Thanks for the info. I look forward to release 0.7. It looks like there will be a lot of improvements. I'll go ahead and close this issue. Thanks.
